main/src/cgeo/geocaching/calculator/ButtonData.java

```diff
@@ -0,0 +1,111 @@
+package cgeo.geocaching.calculator;
+
+import static cgeo.geocaching.models.CalcState.ERROR_CHAR;
+
+import java.io.Serializable;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Data used to capture the state of this particular button such that it can be restored again later
+ */
+public final class ButtonData implements Serializable, JSONAble {
+
+    /** The three states the button can be put into */
+    public enum ValueType {
+        INPUT_VAL {
+            @Override
+            public char getLabel(final ButtonData buttonData) {
+                return buttonData.inputVal;
+            }
+
+            @Override
+            public char setAutoChar(final ButtonData buttonData, final char nextChar) {
+                buttonData.autoChar = nextChar;
+                return nextChar;
+            }
+
+        },
+        AUTO_CHAR {
+            @Override
+            public char getLabel(final ButtonData buttonData) {
+                return buttonData.autoChar;
+            }
+
+            @Override
+            public char setAutoChar(final ButtonData buttonData, final char autoChar) {
+                char nextChar = autoChar;
+                buttonData.autoChar = nextChar++;
+                return nextChar;
+            }
+        },
+        BLANK {
+            @Override
+            public char getLabel(final ButtonData buttonData) {
+                return ButtonData.BLANK;
+            }
+
+            @Override
+            public char setAutoChar(final ButtonData buttonData, final char autoChar) {
+                buttonData.autoChar = autoChar;
+                return autoChar;
+            }
+        },
+        CUSTOM {
+            @Override
+            public char getLabel(final ButtonData buttonData) {
+                return buttonData.customChar;
+            }
+
+            @Override
+            public char setAutoChar(final ButtonData buttonData, final char autoChar) {
+                char nextChar = autoChar;
+                buttonData.autoChar = nextChar;
+                if ('A' <= buttonData.customChar && buttonData.customChar <= 'Z') {
+                    nextChar = buttonData.customChar;
+                    nextChar++;
+                }
+
+                return nextChar;
+            }
+        };
+
+        public abstract char getLabel(ButtonData buttonData);
```

Comment:
I think "getLabel" and "setAutoChar" could be instance methods of class "ButtonData" instead of moving them into the enum. Or am I missing something?

Reply:
I just moved the class out of the CalculatorDialog as whole.
But since there are different types of Buttons, (CustomInput, Blank, AutoChar and CustomChar) which leads to different behaviour, this is handled via the abstract methods getLabel and setAutoChar of the enum